Feature/initial core

zapletin-dk opened the pull request:
Layouts of the framework added, simple test added

zapletin-dk committed:
Framework layers created
Simple login test added

zapletin-dk committed:
Framework layers created
Simple login test added

zapletin-dk committed:
ode refactored

Anassstasi commented on src/main/java/pages/BasePage.java:
```diff
@@ -0,0 +1,14 @@
+    protected AppiumDriver driver;
+
+    public BasePage(AppiumDriver driver) {
+        this.driver = driver;
```
The approach you are using with driver init in the BasePage is absolutely correct BUT i'd recommend to create WebDriverManager where we store driver. It allows us to access driver from any point in the framework without passing it to class constructor from the page plus it helps to handle multithreading in the future (consider that during implementation)

zapletin-dk replied:
Accepted, improved

Anassstasi commented on src/main/java/pages/LoginPage.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public LoginPage fillPasswordField(String password) {
+       // passwordField.click();
```
Please remove reduntant comments

zapletin-dk replied:
Removed

Anassstasi commented on src/main/java/steps/StepDefinitions.java:
```diff
@@ -0,0 +1,30 @@
+        loginPage.allowNotifications()
+                .clickAlreadyHaveAccount()
+                .clickContinueWithEmail()
+                .fillEmailField("yazio_test@yopmail.com")
```
Please don't hardcode credentials inside POs. Even if it's a known user there might be several of them, so it's better to pass them as an argument so that the method is reusable. I'd recommend to create UserFactory class to manage different users

zapletin-dk replied:
I've created User and UserFactory classes to avoid hardcoded credentials in tests

Anassstasi commented on src/main/java/steps/StepDefinitions.java:
```diff
@@ -0,0 +1,30 @@
+                .clickLoginButton();
+    }
+
+    public boolean welcomeTextOccursOnMainPage() {
```
any method name should start with a verb. If it returns boolean it should start with 'is/are'

zapletin-dk replied:
Refactored methods names

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -0,0 +1,58 @@
+        }
+    }
+
+    public AppiumDriver getDriver() throws Exception {
```
If you have getDriver() method inside DriverManager then why do you create a variable for driver in BasePage class? the whole point about WebDriverManager is to avoid passing driver from one layer to another via constructors

zapletin-dk replied:
Refactored

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -0,0 +1,58 @@
+        return INSTANCE;
+    }
+
+    public AppiumDriver initDriver() throws Exception {
```
Please pass driver capabilities as an argument. We expect runing the tests against different devices and we should be able to reuse the methods

Anassstasi commented on src/main/resources/configs/emulator.properties:
```diff
@@ -0,0 +1,7 @@
+automationName=UIAutomator2
+appPackage=com.yazio.android
+appActivity=com.yazio.android/yazio.feature.MainActivity
+remoteURL=http://127.0.0.1:4723/
```
This is not an emulatore URL. That's APPIUM server url . It doesn't belong to this file

Anassstasi replied:
If we have multiple devices, can we use the same Appium server Url?

zapletin-dk replied:
Put in separated environment configuration

Anassstasi commented on src/main/resources/configs/emulator.properties:
```diff
@@ -0,0 +1,7 @@
+deviceName=Pixel_9_android_15
+platformName=android
+version=15.0
+automationName=UIAutomator2
```
Is automationName different for every emulator?

zapletin-dk replied:
No, included automation name directly in DriverManager class

Anassstasi commented on src/test/java/BaseTest.java:
```diff
@@ -0,0 +1,22 @@
+
+public class BaseTest {
+    protected AndroidDriver driver;
+    DriverManager driverManager = DriverManager.getInstance();
```
Why do you need to create an instanse of driverManager? the whole point of it is to be able to call driver via DriverManager.getDriver() from any point in the framework. Please let's stick to that flow

zapletin-dk replied:
Removed this logic

Anassstasi commented on src/test/java/BaseTest.java:
```diff
@@ -0,0 +1,22 @@
+public class BaseTest {
+    protected AndroidDriver driver;
+    DriverManager driverManager = DriverManager.getInstance();
+    StepDefinitions stepDefinitions;
```
Please use access modifiers

Anassstasi commented on src/test/java/BaseTest.java:
```diff
@@ -0,0 +1,22 @@
+    @BeforeMethod
+    public void setUp() throws Exception {
+        driver = (AndroidDriver) driverManager.getDriver();
+        stepDefinitions = new StepDefinitions(driver);
```
Do we need to init stepDefinitions before every method? or once for a test run is enough?

zapletin-dk replied:
According to this implementation, initializing stepDefinition class is needed before every test, because it incapulates initializing PO's inside with PageFactory approach using driver instance.

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -0,0 +1,58 @@
+
+    public void quitDriver() {
+        if (driver != null) {
+            driver.quit();
```
Is driver.quit(); enough or should we do smth else as well? After running quit do we still have driver object inside DriverManager or is it null?

zapletin-dk replied:
Refactored code to assign driver null value after quit()

Anassstasi commented on src/test/java/LoginPageTest.java:
```diff
@@ -0,0 +1,11 @@
+    @Test
+    public void signInAsExistingUserTest(){
+        stepDefinitions.loginAsExistingUser();
+        Assert.assertTrue(stepDefinitions.welcomeTextOccursOnMainPage());
```
Please add a detailed assertion message

zapletin-dk replied:
Added

Anassstasi commented on src/test/java/LoginPageTest.java:
```diff
@@ -0,0 +1,11 @@
+
+public class LoginPageTest extends BaseTest {
+
+    @Test
```
Please add description to the test annotation

zapletin-dk replied:
Added

Anassstasi commented on src/test/java/LoginPageTest.java:
```diff
@@ -0,0 +1,11 @@
+public class LoginPageTest extends BaseTest {
+
+    @Test
+    public void signInAsExistingUserTest(){
```
Please fix code styling and formatting for every class

zapletin-dk replied:
Fixed

Anassstasi reviewed (changes requested):
Please fix the comments

zapletin-dk committed:
Code refactored according to Feature/initial core #1 suggestions.
Added implementation for IOS locators.
DriverManager class extended for both android and ios driver.
UserFactory created for User data handling.
IOS Config and properties files added.

Anassstasi commented on src/main/java/configurations/ConfigReader.java:
```diff
@@ -3,5 +3,7 @@
 
 public class ConfigReader {
-    public static final EmulatorConfig emulatorConfig = ConfigFactory.create(EmulatorConfig.class, System.getProperties());
+    public static final AndroidConfig ANDROID_CONFIG = ConfigFactory.create(AndroidConfig.class, System.getProperties());
```
Why do you need separate configs for iOS and Android? Do the fields which are not common there chnage at all? what's the purpose of this config files?

Anassstasi commented on src/main/java/configurations/IosConfig.java:
```diff
@@ -0,0 +1,19 @@
+})
+public interface IosConfig extends Config {
+    @Key("deviceName")
+    String deviceName();
```
Aren't all these parameters the same for android?

Anassstasi commented on src/main/java/pages/BasePage.java:
```diff
@@ -3,12 +3,15 @@
-        PageFactory.initElements(new AppiumFieldDecorator(driver), this);
+    public BasePage() {
+        try {
+            AppiumDriver driver = DriverManager.getDriver();
```
Why do you need a variable for driver? From what i can see it's used just once.

Anassstasi commented on src/main/java/steps/StepDefinitions.java:
```diff
@@ -1,30 +1,28 @@
     }
 
-    public void loginAsExistingUser() {
+    public void loginAsExistingUser(User user) {
```
In this case we can rename the method to loginAs(...) and it will be more reusable

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
-        if (INSTANCE == null) {
-            INSTANCE = new DriverManager();
+    public static AppiumDriver getDriver() {
+        AppiumDriver driver = driverThreadLocal.get();
```
I would recommend to get rid of extra variable definitions when it comes to driver. Extra points to think about thread safety so it's better to avoid such cases

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
-            INSTANCE = new DriverManager();
+    public static AppiumDriver getDriver() {
+        AppiumDriver driver = driverThreadLocal.get();
+        if (driver == null) {
```
riverThreadLocal.get() == null

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
+        AppiumDriver driver = driverThreadLocal.get();
+        if (driver == null) {
+            driver = initDriver();
+            driverThreadLocal.set(driver);
```
set(initDriver())

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
-            driver.quit();
+    private static DriverFactory createDriverFactory() {
+        String platformName = ConfigReader.ENVIRONMENT_CONFIG.os().toLowerCase();
+        if (platformName.contains("android")) {
```
You will have to check if it's android or ios quite often so id recommend to create a util class for that

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
-    public void quitDriver() {
-        if (driver != null) {
-            driver.quit();
+    private static DriverFactory createDriverFactory() {
```
What's the Factory pattern? 
what's new AndroidDriverFactory(); and why can't it be factory?

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
+            try {
+                driver.quit();
+            } finally {
+                driverThreadLocal.remove();
```
why have u put it inside finally? in my opinion it's not necessary

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
+
+    private enum RunMode {
+        LOCAL("local"),
+        REMOTE("remote", "http://remote-appium-server:4723"),
```
What's the difference between remote and ci?

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
+        }
+    }
+
+    private static class IOSDriverFactory implements DriverFactory {
```
This method is not a factory. Please check factory pattern.

DriverFactory is factory because it created the same object (driver) of different types (ios/android) depending on OS type

Anassstasi commented on src/main/java/utils/UserFactory.java:
```diff
@@ -0,0 +1,20 @@
+import com.github.javafaker.Faker;
+
+public class UserFactory {
+    public static User existingUser;
```
Why is it public?

zapletin-dk replied:
Removed

Anassstasi commented on src/main/java/utils/UserFactory.java:
```diff
@@ -0,0 +1,20 @@
+
+import com.github.javafaker.Faker;
+
+public class UserFactory {
```
The whole idea of this class is to store/manage credentials and users. You don't need any static blocks for that. You can just create getExistingUser() method and return existing user object there.

Anassstasi commented on src/main/resources/configs/ios.properties:
```diff
@@ -0,0 +1,4 @@
+deviceName=Iphone 12
```
Instead of ios properties i would name it iosDevice properties or smth like that. Ios properties are static and never change while device properties change. Keep in mind you might have parallel execution against different devices

Anassstasi replied:
Same for android

Anassstasi commented on src/test/java/BaseTest.java:
```diff
@@ -1,22 +1,18 @@
-        driver = (AndroidDriver) driverManager.getDriver();
-        stepDefinitions = new StepDefinitions(driver);
+    public void setUp() {
+        stepDefinitions = new StepDefinitions();
```
Do u need to init stepDefs before each method? Do they change somehow from method to method?

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
-        driver = new AndroidDriver(new URI(URL).toURL(), capabilities);
-        driver.manage().timeouts().implicitlyWait(Duration.ofSeconds(20));
+        DriverFactory driverFactory = createDriverFactory();
+        AppiumDriver driver = driverFactory.createDriver(appiumUrl, capabilities);
```
To create driver the only parameter that matters for you is OS, the rest is more or less static. (keep in mind device info)

Anassstasi commented on src/main/java/utils/DriverManager.java:
```diff
@@ -3,56 +3,127 @@
     }
 
-    public AppiumDriver initDriver() throws Exception {
+    private static AppiumDriver initDriver() {
```
i would recommend to pass driver info as an argument  to this method, it should be clear and transparent what driver you are creating while with current implementation - it's not.

zapletin-dk committed:
Core refactored
Builder and lombok implemented for User.class
DriverFactory class created
Configurations should be replaced with TestNG xml in the next iteration

zapletin-dk commented on src/main/java/configurations/DriverManager.java:
```diff
@@ -0,0 +1,32 @@
+
+    public static AppiumDriver getDriver() {
+        if (driverThreadLocal.get() == null) {
+            driverThreadLocal.set(initDriver("ANDROID"));
```
This is temporary usage, TestNG xml logic will be implemented in the next iterations

zapletin-dk committed:
TestNG approach implemented
Unnecessary configurations removed
Driver init method moved to DriverFactory class during to logic

zapletin-dk committed:
README file added
Pom.xml cleaned

zapletin-dk committed:
Update README.md

Anassstasi commented on pom.xml:
```diff
@@ -0,0 +1,85 @@
+            <artifactId>log4j-core</artifactId>
+            <version>${log4j.version}</version>
+        </dependency>
+        <dependency>
```
Add a space line between these dependencies

zapletin-dk replied:
Added

Anassstasi commented on src/main/java/entity/User.java:
```diff
@@ -0,0 +1,52 @@
+    private String email;
+    private String phone;
+
+    private User(Builder builder) {
```
Lombok has a built-in builder but let's keep the implementation that from scratch anyway. This is JFYI

zapletin-dk replied:
Sure

Anassstasi commented on src/main/java/entity/UserFactory.java:
```diff
@@ -0,0 +1,24 @@
+
+public class UserFactory {
+
+    Faker faker = new Faker();
```
Please add access modifier

zapletin-dk replied:
Added modifier and made variable static

Anassstasi commented on src/main/java/pages/BasePage.java:
```diff
@@ -0,0 +1,15 @@
+
+abstract class BasePage {
+    public BasePage() {
+        try {
```
Why do u need try/catch here? Don't think it's necessary

zapletin-dk replied:
Agree, this is redundant now. This  approach helped me to handle error only here and stop throwing exception in previous PO initialization approach.

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+        DriverManager.setDriverThreadLocal(driver);
+    }
+
+    private static AppiumDriver createDriver(String device, String deviceOS) {
```
Let's create a enum for platfirms

zapletin-dk replied:
Done

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+        String appiumUrl = getAppiumUrl();
+
+        if (deviceOS.toLowerCase().contains("android")) {
+            capabilities.setCapability("deviceName", device);
```
Please create a method for android driver config and reuse it. Same for iOS

zapletin-dk replied:
Done

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+            capabilities.setCapability("appPackage", "com.yazio.android");
+            capabilities.setCapability("appActivity", "com.yazio.android/yazio.feature.MainActivity");
+            capabilities.setCapability("automationName", "UiAutomator2");
+            try {
```
Is try/catch that necessary here? Don't think it is. Same for ios

zapletin-dk replied:
Try catch is necessary because of handling UriStntaxExceotion and Unknow os name exception

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+            this(mode, null);
+        }
+
+        RunMode(String mode, String defaultUrl) {
```
Why do u need defaultUrl? can't see any option with that parameter

zapletin-dk replied:
Removed

Anassstasi commented on src/test/java/BaseTest.java:
```diff
@@ -0,0 +1,22 @@
+    @Parameters({"device", "deviceOS"})
+    public void setUp(String device, String deviceOS) {
+        DriverFactory.initDriver(device, deviceOS);
+        stepDefinitions = new StepDefinitions();
```
Why do u init stepDefinitions before every method? it's not necessary to do that so often.

Anassstasi commented on src/test/java/LoginPageTest.java:
```diff
@@ -0,0 +1,14 @@
+    @Description("Test to verify that the user can sign in with an existing account")
+    public void signInAsExistingUserTest() {
+        stepDefinitions.loginAs(UserFactory.getExistingUser());
+        Assert.assertTrue(stepDefinitions.isWelcomeBackTextVisibleAfterLogIn(), "Welcome back text is visible after login");
```
Please check assertion message. It appears if the assertion fails but for some reason it says 'Welcome back text is visible after login' so it seems that we are expecting it to not be visible

zapletin-dk replied:
Fixed message

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+package utils;
```
DriverFacctory is not a Utils class. It's part of configuration, isn't it?

zapletin-dk replied:
Moved to Configuration package

Anassstasi commented on src/test/java/resources/android_suite.xml:
```diff
@@ -0,0 +1,10 @@
+<!DOCTYPE suite SYSTEM "https://testng.org/testng-1.0.dtd">
+<suite name="Android_Suite" verbose="2">
+    <test name="AndroidTest">
+        <parameter name="device" value="Pixel_9_android_15"/>
```
For configuration OS version is necessary as well, can't see it mentioned anywhere

zapletin-dk replied:
added

Anassstasi commented on src/test/java/resources/ios_suite.xml:
```diff
@@ -0,0 +1,10 @@
+<suite name="Ios_Suite" verbose="2">
+    <test name="IOSTest">
+        <parameter name="device" value="iPhone 12"/>
+        <parameter name="deviceOS" value="ios"/>
```
Same comment about OS version

Anassstasi commented on src/main/java/utils/DriverFactory.java:
```diff
@@ -0,0 +1,84 @@
+
+    public static void initDriver(String device, String deviceOS) {
+        AppiumDriver driver = DriverFactory.createDriver(device, deviceOS);
+        driver.manage().timeouts().implicitlyWait(IMPLICIT_WAIT_TIMEOUT);
```
Please move implicitlyWait to createDriver method as it's part of driver creation not initialization

zapletin-dk replied:
Moved

Anassstasi reviewed (changes requested):
PLease resolve the comments

zapletin-dk committed:
Fixes and code clean-up

zapletin-dk committed:
Deleted unnecessary test
